src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/transforms/charts/legacy_metric.ts

```diff
@@ -0,0 +1,205 @@
+import type { LensAttributes } from '../../types';
+import { DEFAULT_LAYER_ID } from '../../types';
+import {
+  addLayerColumn,
+  buildDatasetState,
+  buildDatasourceStates,
+  buildReferences,
+  generateApiLayer,
+  getAdhocDataviews,
+  operationFromColumn,
+} from '../utils';
+import { getValueApiColumn, getValueColumn } from '../columns/esql_column';
+import type { LensApiState, LegacyMetricState } from '../../schema';
+import { fromMetricAPItoLensState } from '../columns/metric';
+import type { LensApiAllMetricOperations } from '../../schema/metric_ops';
+import type { DeepMutable, DeepPartial } from '../utils';
+import { generateLayer } from '../utils';
+import type {
+  LegacyMetricStateESQL,
+  LegacyMetricStateNoESQL,
+} from '../../schema/charts/legacy_metric';
+import { getSharedChartLensStateToAPI, getSharedChartAPIToLensState } from './utils';
+import { fromColorByValueAPIToLensState, fromColorByValueLensStateToAPI } from '../coloring';
+
+const ACCESSOR = 'metric_formula_accessor';
+const LENS_DEFAULT_LAYER_ID = 'layer_0';
+
+function buildVisualizationState(config: LegacyMetricState): LegacyMetricVisualizationState {
+  const layer = config;
+
+  return {
+    layerId: DEFAULT_LAYER_ID,
+    layerType: 'data',
+    accessor: ACCESSOR,
+    size: layer.metric.size,
+    titlePosition: layer.metric.alignments?.labels,
+    textAlign: layer.metric.alignments?.value,
+    ...(layer.metric.apply_color_to && layer.metric.color
+      ? {
+          colorMode: layer.metric.apply_color_to === 'background' ? 'Background' : 'Labels',
+          palette: fromColorByValueAPIToLensState(layer.metric.color),
+        }
+      : { colorMode: 'None' }),
+  };
+}
+
+function reverseBuildVisualizationState(
+  visualization: LegacyMetricVisualizationState,
+  layer: FormBasedLayer | TextBasedLayer,
+  layerId: string,
+  adHocDataViews: Record<string, DataViewSpec>,
+  references: SavedObjectReference[],
+  adhocReferences?: SavedObjectReference[]
+): LegacyMetricState {
+  if (visualization.accessor === undefined) {
+    throw new Error('Metric accessor is missing in the visualization state');
+  }
+
+  const dataset = buildDatasetState(layer, adHocDataViews, references, adhocReferences, layerId);
+
+  if (!dataset || dataset.type == null) {
+    throw new Error('Unsupported dataset type');
+  }
+
+  let props: DeepPartial<DeepMutable<LegacyMetricState>> = generateApiLayer(layer);
+
+  if (dataset.type === 'esql' || dataset.type === 'table') {
+    const esqlLayer = layer as TextBasedLayer;
+    props = {
+      ...props,
+      metric: getValueApiColumn(visualization.accessor, esqlLayer),
+    } as LegacyMetricState;
+  } else if (dataset.type === 'dataView' || dataset.type === 'index') {
+    const formLayer = layer as FormBasedLayer;
+    const metric = operationFromColumn(
+      visualization.accessor,
+      formLayer
+    ) as LensApiAllMetricOperations;
+
+    props = {
+      ...props,
+      metric,
+    } as LegacyMetricState;
+  }
```

Comment:
I wish we would use a nicer pattern here.
Something like:


```suggestion
  const props: LegacyMetricState = {
  ...generateApiLayer(layer),
  metric: isESQLDataset(dataset) ? getValueApiColumn(visualization.accessor, layer as TextBasedLayer) : operationFromColumn(
      visualization.accessor,
      layer as FormBasedLayer
    )
  }
```